File: python/Lec 2 Build-up Logical Thinking (Patterns)/pattern_07.py
#     *
#    ***
#   *****
#  *******
# *********


#----------------------------------
#  BRUTE FORCE APPROACH
# ----------------------------------
n = 5
for i in range(1, n + 1):
    line = ""
    # Add spaces
    for j in range(n - i):
        line += " "  # O(n²) string copying
    # Add stars
    for k in range(2 * i - 1):
        line += "*"
    print(line)

# ----------------------------------
# BETTER APPROACH (String multiplication)
# ----------------------------------

n = 5
for i in range(1, n + 1):
    spaces = " " * (n - i)
    stars = "*" * (2 * i - 1)
    print(spaces + stars)

# ----------------------------------
#  BEST APPROACH (List comprehension)
# ----------------------------------
n = 5
print('\n'.join([
    " " * (n - i) + "*" * (2 * i - 1) 
    for i in range(1, n + 1)
]))

File: cpp/Lec 2 Build-up Logical Thinking (Patterns)/pattern_07.cpp
//     *
//    ***
//   *****
//  *******
// *********

//----------------------------------
// BRUTE FORCE APPROACH (Nested loops)
//----------------------------------
#include <iostream>
using namespace std;
int main()
{
    int n = 5;
    for (int i = 1; i <= n; i++)
    {
        // Print spaces
        for (int j = 1; j <= n - i; j++)
        {
            cout << " ";
        }
        // Print stars
        for (int k = 1; k <= 2 * i - 1; k++)
        {
            cout << "*";
        }
        cout << endl;
    }
    return 0;
}

//----------------------------------
// BETTER APPROACH (String constructors)
//----------------------------------
#include <iostream>
#include <string>
using namespace std;
int main()
{
    int n = 5;
    for (int i = 1; i <= n; i++)
    {
        string spaces(n - i, ' ');
        string stars(2 * i - 1, '*');
        cout << spaces << stars << endl;
    }
    return 0;
}

//----------------------------------
// BEST APPROACH (Single pre-allocated string)
//----------------------------------
#include <iostream>
#include <string>
using namespace std;
int main()
{
    int n = 5;
    string line(n + 4, ' '); // Maximum width: n-1 spaces + (2*n-1) stars

    for (int i = 1; i <= n; i++)
    {
        int spaces = n - i;
        int stars = 2 * i - 1;

        // Fill line efficiently
        fill(line.begin(), line.begin() + spaces, ' ');
        fill(line.begin() + spaces, line.begin() + spaces + stars, '*');
        line[spaces + stars] = '\0'; // Null terminate for puts

        puts(line.c_str());
    }
    return 0;
}